Staggered (Variable batch) Updates

rdelval opened the pull request:
Adding support for variable group sizes when executing an update.

Design doc for this change is here: 
https://docs.google.com/document/d/1xGk4ueH8YlmJCk6hQJh85u4to4M1VQD0l630IOchvgY/edit#heading=h.lg3hty82f5cz

Testing done:
* Jenkins integration test passed.
* End to end tests passed.
* Started an update, stopped the aurora-scheduler, checkout out 0.21.0, vagrant ssh'd, aurorabuild all, monitored job.
* Started an update and restarted the aurora-scheduler at random intervals between 3 and 13 seconds. Verified group size remained correct.

Feedback from https://reviews.apache.org/r/66192/ addressed. 

UI Change below:
![image](https://user-images.githubusercontent.com/1643994/45329549-d22e7200-b515-11e8-92d9-eefc31b793b8.png)
![image](https://user-images.githubusercontent.com/1643994/45329562-deb2ca80-b515-11e8-9b76-10ca1945b7f7.png)
![image](https://user-images.githubusercontent.com/1643994/45329565-e4101500-b515-11e8-8a97-bb904bf77ceb.png)

Tests added for UI, Thrift change, and Pystachio changes.

Moving this over to Github since reviewboard decided to act up.

StephanErb commented on src/main/java/org/apache/aurora/scheduler/storage/durability/ThriftBackfill.java:
```diff
@@ -143,8 +148,28 @@ public IJobUpdate backFillJobUpdate(JobUpdate update) {
     return IJobUpdate.build(update);
   }
+
+  private void backfillUpdateStrategy(JobUpdateSettings settings) {
```
I think we should also perform the reverse operation here if possible: If only the new settings are set, also write the deprecated ones.

This does not add much complexity but makes scheduler upgrade/downgrade paths easier to reason about.

rdelval replied:
Will do this. In order to do this, I'll have a lossy conversion from the variable batch update strategy to a static batch update strategy which uses index 0 of the list contained by the variable batch update strategy.

e.g.: VarBatchUpdate = [1,2,3,4] will become BatchUpdate, group Size = 1 if the scheduler is downgraded to 0.19.0.

DavidMcLaughlin replied:
It is more complicated than that. As soon as we persist one of these structs, we will be unable to roll back to any previous version of the software that does not have the thrift changes. To enable safe roll-back/roll-forward, we will need a commit/release with just the Thrift changes.

rdelval replied:
Thanks for the input @DavidMcLaughlin, I'll create a PR with just the Thrift changes in that case.

rdelval replied:
Tested this out by starting a vagrant box. Changed the code inside of [SchedulerThriftInterface.java](https://github.com/rdelval/aurora/blob/staggeredUpdates/src/main/java/org/apache/aurora/scheduler/thrift/SchedulerThriftInterface.java#L798) to set these fields upon receiving the request:

```    // Keep old job schema in case we want to revert to a lower version of Aurora that doesn't
    // support variable update group sizes
    int totalInstancesFromGroups;
    if (settings.getUpdateStrategy().isSetQueueStrategy()) {
      totalInstancesFromGroups = settings.getUpdateStrategy().getQueueStrategy().getGroupSize();
      settings.setWaitForBatchCompletion(false);
      settings.setUpdateGroupSize(settings.getUpdateStrategy().getQueueStrategy().getGroupSize());
    } else if (settings.getUpdateStrategy().isSetBatchStrategy()) {
      totalInstancesFromGroups = settings.getUpdateStrategy().getBatchStrategy().getGroupSize();
      settings.setWaitForBatchCompletion(true);
      settings.setUpdateGroupSize(settings.getUpdateStrategy().getBatchStrategy().getGroupSize());
    } else if (settings.getUpdateStrategy().isSetVarBatchStrategy()) {
      VariableBatchJobUpdateStrategy strategy = settings.getUpdateStrategy().getVarBatchStrategy();
      settings.setWaitForBatchCompletion(true);
      // Setting this value to allow graceful rollback to version 0.21.0
      // TODO(rdelvalle): Remove compatibility settings once version 0.22.0 ships
      settings.setUpdateGroupSize(1); // Necessary because we haven't validated fields yet.
```

I created a Variable batch update, then shut down the scheduler, and checked out tag rel/0.21.0 and ran aurorabuild all on the vagrant box. 

The scheduler came up fine. I forced a snapshot and a backup, restarted the scheduler with out issue. I upgraded the vagrant box back to the staggered updates patch and there was info loss (update no longer had staggered updates fields set so it stayed as a batch update), which in this case is an acceptable outcome IMO. 

Will update the patch with all the changes.

StephanErb commented on src/main/java/org/apache/aurora/scheduler/updater/strategy/VariableBatchStrategy.java:
```diff
@@ -0,0 +1,135 @@
+      sum = maxActiveGroups.get(step);
+
+      // TODO(rdelvalle): Consider if it's necessary to handle fractional steps.
+      // (i.e. halfway between step X and X+1)
```
I think now would be the best time to reason about this. I don't think fractional steps are necessary, but maybe you can elaborate what you had in mind.

rdelval replied:
This thought may have come up as a side effect of the old way of calculating the current batch size. What I was thinking about is if half way through a batch size, the scheduler fails over and the update has to continue from where it left off when the next leader is chosen.

E.g: We have batch sizes 10 and 30. Half way through 30 (25 instances updated), the scheduler fails over.

With the new way of calculating which batch size we're on on the fly, I don't think it's an issue anymore.

StephanErb commented on src/main/java/org/apache/aurora/scheduler/updater/strategy/VariableBatchStrategy.java:
```diff
@@ -0,0 +1,135 @@
+    this.instanceModificationCount = Optional.empty();
+  }
+
+  private int determineStep(int idle) {
```
Please add one short comment that explains what this function does. For a casual reader it can be confusing what you mean with step.

rdelval replied:
I agree, I think this function would actually benefit more from having it's name changed to `determineCurBatchSize`.

rdelval commented on docs/reference/configuration.md:
```diff
@@ -376,8 +376,34 @@ Parameters for controlling the rate and policy of rolling updates.
+### VariableBatchUpdateStrategy Objects
+
+Similar to Batch Update strategy, this strategy will wait until all instances in a current group are updated before updating more instances.
+However, with this strategy the size of each group may change as the update progresses.
```
Good catch!

rdelval commented:
@StephanErb updated the PR to address feedback. Thank you very much for the feedback, it helped me realize rollback was not exactly a mirror image of roll forward. I've addressed that with this update.

rdelval commented on src/main/java/org/apache/aurora/scheduler/thrift/SchedulerThriftInterface.java:
```diff
@@ -792,9 +796,51 @@ public Response startJobUpdate(JobUpdateRequest mutableRequest, @Nullable String
+
+    // Gracefully handle a client sending an update with an older thrift schema
+    // TODO(rdelvalle): Remove after version 0.22.0 ships
+    if (!settings.isSetUpdateStrategy()) {
```
Agreed, fixed.

rdelval commented on src/main/java/org/apache/aurora/scheduler/thrift/SchedulerThriftInterface.java:
```diff
@@ -792,9 +796,51 @@ public Response startJobUpdate(JobUpdateRequest mutableRequest, @Nullable String
+      }
+    }
+
+    // Keep old job schema in case we want to revert to a lower version of Aurora that doesn't
```
That's actually a great point. I removed the code, ran the code, generated a variable batch update, and downgraded. Scheduler came up just fine.

jordanly replied:
Just a thought: is it still possible to create an update outside of the CLI (calling the API directly) without these fields (or setting them as null)? If so, new updates will not have this field set so they could crash the scheduler on a downgrade.

I don't have strong feelings on this since it is probably unlikely, but wanted to alert to the possibility.

rdelval commented on src/main/python/apache/aurora/config/schema/base.py:
```diff
@@ -38,6 +46,7 @@ class UpdateConfig(Struct):
   wait_for_batch_completion   = Default(Boolean, False)
   pulse_interval_secs         = Integer
   sla_aware                   = Default(Boolean, False)
+  update_strategy = Choice([QueueUpdateStrategy, BatchUpdateStrategy, VariableBatchUpdateStrategy])
```
I think I fixed this with the latest push but let me know if the style is still not right.

rdelval commented:
@wfarner any chance you could take a quick look at this patch?

wfarner commented:
@rdelval sorry, this is too large for me to commit to, and i wouldn't want to interfere with the existing review activity.

rdelval commented:
No worries, I agree, the patch is quite big at this point. I've considered breaking up this patch into multiple patches if that'll make it easier to review. @StephanErb and @DavidMcLaughlin would a few smaller patches be preferable to one large one?

shanmugh commented on src/main/python/apache/aurora/config/schema/base.py:
```diff
@@ -38,6 +46,9 @@ class UpdateConfig(Struct):
   wait_for_batch_completion   = Default(Boolean, False)
   pulse_interval_secs         = Integer
   sla_aware                   = Default(Boolean, False)
+  update_strategy = Choice([QueueUpdateStrategy,
```
style nit - tabs to align

shanmugh commented on src/test/python/apache/aurora/client/api/test_updater_util.py:
```diff
@@ -46,15 +71,94 @@ def test_empty_list(self):
+  def test_to_thrift_update_settings_strategy(self):
+
+    """Test to_thrift produces an expected thrift update settings configuration
+       from a PyStachio update object.
```
nit - `s/PyStachio/Pystachio/`

shanmugh commented on src/test/python/apache/aurora/client/api/test_updater_util.py:
```diff
@@ -46,15 +71,94 @@ def test_empty_list(self):
+  def test_to_thrift_update_settings_no_strategy_queue(self):
+
+    """Test to_thrift produces an expected thrift update settings configuration
+       from a PyStachio update object that doesn't include an update strategy.
```
nit - `s/PyStachio/Pystachio/`

shanmugh commented on src/test/python/apache/aurora/client/api/test_updater_util.py:
```diff
@@ -46,15 +71,94 @@ def test_empty_list(self):
+  def test_to_thrift_update_settings_no_strategy_batch(self):
+
+    """Test to_thrift produces an expected thrift update settings configuration
+       from a PyStachio update object that doesn't include an update strategy.
```
nit - `s/PyStachio/Pystachio/`

jordanly commented on src/main/java/org/apache/aurora/scheduler/thrift/SchedulerThriftInterface.java:
```diff
@@ -1026,7 +1051,6 @@ private static IJobUpdateKey validateJobUpdateKey(JobUpdateKey mutableKey) {
   static String noCronScheduleMessage(IJobKey jobKey) {
     return String.format("Job %s has no cron schedule", JobKeys.canonicalString(jobKey));
   }
-
```
Keep whitespace

jordanly commented on src/main/java/org/apache/aurora/scheduler/updater/strategy/VariableBatchStrategy.java:
```diff
@@ -0,0 +1,152 @@
+   * @param maxActiveGroups  List of Maximum group sizes. Each group size represents a step.
+   * {@link #getNextGroup(Set, Set)}.
+   */
+  public VariableBatchStrategy(Ordering<T> ordering,
```
nit: ordering should be on its own line

jordanly commented on src/main/java/org/apache/aurora/scheduler/updater/strategy/VariableBatchStrategy.java:
```diff
@@ -0,0 +1,152 @@
+  // Determine how far we're into the update based upon how many instances are waiting
+  // to be modified.
+  private int determineCurGroupSize(int remaining) {
+
```
nit: remove extra newline

jordanly commented on src/main/java/org/apache/aurora/scheduler/updater/strategy/VariableBatchStrategy.java:
```diff
@@ -0,0 +1,152 @@
+        totalModInstanceCount.get());
+
+    if (rollingForward) {
+
```
nit: remove extra newline

jordanly commented on src/main/java/org/apache/aurora/scheduler/updater/strategy/VariableBatchStrategy.java:
```diff
@@ -0,0 +1,152 @@
+      // Return last step when number of instances > sum of all groups
+      return lastGroupSize;
+    } else {
+
```
nit: remove extra newline

jordanly commented on src/main/java/org/apache/aurora/scheduler/updater/strategy/VariableBatchStrategy.java:
```diff
@@ -0,0 +1,152 @@
+
+  @Override
+  public final Set<T> getNextGroup(Set<T> idle, Set<T> active) {
+
```
nit: remove extra newline

jordanly commented on src/main/java/org/apache/aurora/scheduler/updater/strategy/VariableBatchStrategy.java:
```diff
@@ -0,0 +1,152 @@
+   * @return A subset of {@code idle}, instances to start updating.
+   */
+  Set<T> doGetNextGroup(Set<T> idle, Set<T> active) {
+    if (active.isEmpty()) {
```
I would copy the current `BatchStrategy` impl of this since it is the same:
`return active.isEmpty() ? idle : ImmutableSet.of();`

jordanly commented on src/main/java/org/apache/aurora/scheduler/updater/strategy/VariableBatchStrategy.java:
```diff
@@ -0,0 +1,152 @@
+  @Override
+  public final Set<T> getNextGroup(Set<T> idle, Set<T> active) {
+
+    // Get the size for the idle set on the first run only. This is representative of the number
```
What is the behavior on scheduler restart? Is `totalModInstanceCount` smaller than at the beginning of the update depending on how far the update has gotten before the failover?

rdelval replied:
This was one of my biggest concerns as well. From actually running the code I found the `totalModInstanceCount` always remains the same value.

Our update mechanism is a bit tough to grok but if I'm understanding it correctly, it seems like every time you encounter an update, the whole process is played out from beginning to end. 

This, combined with the fact that when you try apply an update to an already updated instance results in a noop, makes for a nice catch all process where we can avoid worrying about corner cases (how do we recover our state without writing it to the storage).

Same applies to a recovery. So in the end you always end up with that first idle group being the total amount of instances you will modify by the end of the update.

jordanly replied:
Appreciate the rigorous testing! Looking at the code as well convinces me that this logic is fine.

jordanly commented on src/main/java/org/apache/aurora/scheduler/updater/strategy/VariableBatchStrategy.java:
```diff
@@ -0,0 +1,152 @@
+        }
+      }
+      // Return last step when number of instances > sum of all groups
+      return lastGroupSize;
```
`Iterables.getLast` is another alternative from doing your own `.get(size() - 1)`

rdelval commented on src/main/python/apache/aurora/client/api/updater_util.py:
```diff
@@ -80,6 +84,18 @@ def to_thrift_update_settings(self, instances=None):
     instances - optional list of instances to update.
     """
+
+    if self.update_strategy is Empty:
```
This would be a good backwards compatibility guard. I'll go ahead and add that.

rdelval replied:
Added a check for these two scenarios as well as tests for them. If you can, please do a sanity check on these.

rdelval commented on src/test/sh/org/apache/aurora/e2e/test_end_to_end.sh:
```diff
@@ -364,7 +364,6 @@ test_update() {
   assert_update_state_by_id $_jobkey $_update_id 'ROLLED_FORWARD'
 }
-
 test_update_fail() {
```
The variable batch is being tested [here](https://github.com/apache/aurora/pull/37/files/49646d0097546cc070fb0a4b3555cca482dceb28#diff-bc249f48d9141c06b10bfd6a9c8c9ed1R980)

If you mean whether I'm verifying if each batch has the correct number of activate instances, then in this case no. It would be a good addition (do we have such a check for static batch updates already?).

shanmugh replied:
The second one. I don't believe we have a check for static batch strategy. Will be good to expand on it.

rdelval replied:
Left a TODO on the end to end tests file. If it it's fine with you, I'll get to this in a separate PR.

shanmugh commented on src/main/java/org/apache/aurora/scheduler/updater/UpdateFactory.java:
```diff
@@ -76,12 +77,10 @@ Update newUpdate(
       checkArgument(
           settings.getMinWaitInInstanceRunningMs() >= 0,
           "Min wait in running must be non-negative.");
-      checkArgument(
```
I haven't dove into the history of why this line of code was added. But keeping it similar sounds safe to me.

rdelval commented:
@jordanly @shanmugh addressed all feedback. Thank you both very much for the reviews and suggestions. Please do a sanity check on these recent changes. If all goes well let's ship this!
